glos_qartod/run.py

```diff
@@ -45,26 +47,48 @@ def qc_subset(dir_root, conf, mappings):
-                for nc_file in (f for f in fname if f.endswith('.nc')):
-                    # TODO: Add some sort of caching check.
-                    file_dest = os.path.join(root, nc_file)
-                    # check if all the variables already exist in the file
-                    ds = Dataset(file_dest)
-                    # if all of the variables for the expected QC variables
-                    # aren't present in the dataset, add the file to the list
-                    # of files to be QCed
-                    # sometimes naming conventions are inconsistent, so
-                    # check both the remapped version and the regular version
-                    if not (qc_varnames.issubset(ds.variables) or
-                            qc_varnames_bkp.issubset(ds.variables)):
-                        files.append(file_dest)
-        else:
-            # TODO: Add logging noting nonexistent directory
-            continue
+        # get the directory matching this station name or get all if * glob
+        # is used
+        station_dirs = glob.glob(os.path.join(dir_root, var_dir, station))
+        for dest_dir in station_dirs:
+            files.extend(find_files(dest_dir, qc_varnames, qc_varnames_bkp))
+
+    return set(files)
+
+
+def find_files(dest_dir, qc_varnames, qc_varnames_bkp):
+    files = []
+    if os.path.exists(dest_dir):
+        for root, subdir, fname in os.walk(dest_dir):
+            for nc_file in (f for f in fname if f.endswith('.nc')):
+                file_dest = os.path.join(root, nc_file)
+                # check if all the variables already exist in the file
+                # if all of the variables for the expected QC variables
+                # aren't present in the dataset, add the file to the list
+                # of files to be QCed
+
+                qc_filepath = file_dest.rsplit('.', 1)[0] + '.ncq'
+                # try to fetch the QC file's variable names.  If it does not
+                # exist, no QC has been applied and it must be created later
+                if os.path.exists(qc_filepath):
+                    try:
+                        with Dataset(qc_filepath) as f:
+                            qc_vars = f.variables.keys()
+                    # if for some reason we can't open the file,
+                    # note the exception and treat the qc variables as empty
+                    except Exception as e:
+                        get_logger().error('Failed to open file %s: %s',
+                                           qc_filepath, str(e))
+                        qc_vars = []
+                else:
+                    qc_vars = []
+                # check if all the QC variables exist in the file.
+                # if they don't, add them to the list of files to be processed
+                if not (qc_varnames.issubset(qc_vars) or
+                        qc_varnames_bkp.issubset(qc_vars)):
+                    files.append(file_dest)
```

Comment:
The inner body of the inner-most for loop should be a different function, there is too much indentation here.